ocaml-eglot-type-enclosing.el

```diff
@@ -0,0 +1,171 @@
+
+;;; Customizable variables
+
+(defcustom ocaml-eglot-type-buffer-name "*ocaml-eglot-types*"
+  "The name of the buffer storing types."
+  :group 'ocaml-eglot
+  :type 'string)
+
+;;; Internal variables
+
+(defvar-local ocaml-eglot-type-enclosing-types nil
+  "Current list of enclosings related to types.")
+
+(defvar-local ocaml-eglot-type-enclosing-current-type nil
+  "Current type for the current enclosing.")
+
+(defvar-local ocaml-eglot-type-enclosing-offset 0
+  "The offset of the requested enclosings.")
+
+(defvar-local ocaml-eglot-type-enclosing-verbosity nil
+  "The verbosity of the current enclosing request.")
+
+;;; Key mapping for type enclosing
+
+(defvar ocaml-eglot-type-enclosing-map
+  (let ((keymap (make-sparse-keymap)))
+    (define-key keymap (kbd "C-<up>") #'ocaml-eglot-type-enclosing-grow)
+    (define-key keymap (kbd "C-<down>") #'ocaml-eglot-type-enclosing-shrink)
+    (define-key keymap (kbd "C-w") #'ocaml-eglot-type-enclosing-copy)
+    (define-key keymap (kbd "C-c C-t") #'ocaml-eglot-type-enclosing-increase-verbosity)
+    (define-key keymap (kbd "C-<right>") #'ocaml-eglot-type-enclosing-increase-verbosity)
+    (define-key keymap (kbd "C-<left>") #'ocaml-eglot-type-enclosing-decrease-verbosity)
+    keymap)
+  "Keymap for OCaml-eglot's type enclosing transient mode.")
+
+;;; Internal functions
+
+(defun ocaml-eglot-type-enclosing-copy ()
+  "Copy the type of the current enclosing to the Kill-ring."
+  (interactive)
+  (when ocaml-eglot-type-enclosing-current-type
+    (eglot--message "Copied `%s' to kill-ring"
+                    ocaml-eglot-type-enclosing-current-type)
+    (kill-new ocaml-eglot-type-enclosing-current-type)))
+
+(defun ocaml-eglot-type-enclosing--with-fixed-offset ()
+  "Compute the type enclosing for a dedicated offset."
+  (let* ((verbosity ocaml-eglot-type-enclosing-verbosity)
+         (index ocaml-eglot-type-enclosing-offset)
+         (at (ocaml-eglot-util--current-position-or-range))
+         (result (ocaml-eglot-req--type-enclosings at index verbosity))
+         (type (cl-getf result :type)))
+    (setq ocaml-eglot-type-enclosing-current-type type)
+    (ocaml-eglot-type-enclosing--display type t)))
+
+(defun ocaml-eglot-type-enclosing-increase-verbosity ()
+  "Increase the verbosity of the current request."
+  (interactive)
+  (if ocaml-eglot-type-enclosing-verbosity
+      (setq ocaml-eglot-type-enclosing-verbosity
+            (1+ ocaml-eglot-type-enclosing-verbosity))
+    (setq ocaml-eglot-type-enclosing-verbosity 1))
+  (ocaml-eglot-type-enclosing--with-fixed-offset))
+
+(defun ocaml-eglot-type-enclosing-decrease-verbosity ()
+  "Decrease the verbosity of the current request."
+  (interactive)
+  (if ocaml-eglot-type-enclosing-verbosity
+      (progn
+        (setq ocaml-eglot-type-enclosing-verbosity
+              (1- ocaml-eglot-type-enclosing-verbosity))
+        (when (< ocaml-eglot-type-enclosing-verbosity 1)
+          (setq ocaml-eglot-type-enclosing-verbosity nil)))
```

Comment:
Why setting the verbosity twice instead of checking beforehand as you do for growing/shrinking enclosings ?